### ISSUE TITLE
add constraints for port connectors

### ISSUE THREAD

**Issue by minhnh** [rosym-project/algorithm-dsl#5]
- [x] No connection across containers' boundaries allowed, i.e. ports of a block can only be connected to ports of its "siblings" (blocks which have the same container) or ports of its container (scoping)
- [x] A connection between 2 ports should not be repeated (checking rule)
- [x] A `DataConnector` or `TriggerConnector` should not connect a port to itself (checking rule)
- [x] A connector cannot connect ports of the same block (checking rule)
- [x] port direction constraint for ports of parent block vs sibling blocks (checking rule)
- [x] ports of the same block should have unique names (checking rule)
- [x] there should be no unconnected port (checking rule)
- [ ] constraints on whether a data connection should be allowed:

|  data connection to | `FunctionBlock` | `DataBlock` | `SchedulerBlock` |
| :--- | :---: | :---: | :---: |
| "sibling" `FunctionBlock` | n | y | n (?) |
| "sibling" `DataBlock` | y | ? | y (?) |
| "sibling" `SchedulerBlock` | n (?) | y (?) | n (?) |
| containing `FunctionBlockContainer` | y | y | y (?) |
| containing `DataBlockContainer` | - | y | - |

- [ ] constraints on whether a trigger connection should be allowed

|  trigger connection to | `FunctionBlock` | `SchedulerBlock` |
| :--- | :---: | :---: |
| "sibling" `FunctionBlock` | ? | y (?) |
| "sibling" `SchedulerBlock` | y (?) | y (?) |
| containing `FunctionBlockContainer` | y | y |

**Comment by minhnh** [comment]
* `DataConnector` or `TriggerConnector` shouldn't connect ports which both have `In` or `Out` direction

**Comment by svenschneider** [comment]
All those constraints apply for the trigger connections, too.

**Comment by svenschneider** [comment]
We should enforce a unique name/identifier policy on the ports in the same block.

**Comment by svenschneider** [comment]
We should enforce a unique name/identifier policy on the blocks contained in another block.

**Comment by scymtym** [comment]
Can we take some of the necessary constraints from the NPC 4 paper like Herman suggested?

**Comment by svenschneider** [comment]
Of course! It would be nice if we were able to "reuse" the constraints in other BPC-compliant metamodels (e.g. by M2M transformations).

**Comment by minhnh** [comment]
@scymtym could you link to the NPC 4 paper here as well for reference?

**Comment by scymtym** [comment]
> @scymtym could you link to the NPC 4 paper here as well for reference?

The paper can be found here: https://core.ac.uk/display/45288978

**Comment by minhnh** [comment]
I will reopen this issue with for constraints on which data connection to be allowed.
